include/libvfio-user.h

```diff
@@ -897,19 +885,40 @@ bool
 vfu_sg_is_mappable(vfu_ctx_t *vfu_ctx, dma_sg_t *sg);
 
 /*
+ * Creates a new ioeventfd at the given setup memory region with @offset, @size,
+ * @fd, @flags and @datamatch.
+ *
+ * Returns 0 on success and -1 on failure with errno set.
  *
  * @vfu_ctx: the libvfio-user context
+ * @region_idx: The index of the memory region to set up the ioeventfd
+ * @fd: the value of the file descriptor
  * @offset: The offset into the memory region
+ * @size: size of the memory region
  * @flags: Any flags to set up the ioeventfd
- * @index: The index of the memory region to set up the ioeventfd
+ * @datamtach: sets the datamatch value
  */
 int
-create_ioeventfd(vfu_ctx_t *vfu_ctx, size_t offset, uint32_t flags,
-                 uint32_t index);
+vfu_create_ioeventfd(vfu_ctx_t *vfu_ctx, uint32_t region_idx, int fd,
+                     size_t offset, uint32_t size, uint32_t flags,
+                     uint64_t datamatch);
 
-int vfu_create_ioeventfd(vfu_ctx_t *vfu_ctx, size_t offset, uint32_t size, int fd, uint32_t flags, uint32_t index);
-
-int vfu_delete_ioeventfd(vfu_ctx_t *vfu_ctx, uint32_t index, uint32_t fd_index);
+/*
+ * Deletes an ioeventfd at the given setup memory region with @offset, @size,
+ * @fd_index, @flags and @datamatch.
+ *
+ * Returns 0 on success and -1 on failure.
+ *
+ * @vfu_ctx: the libvfio-user context
+ * @region_idx: The index of the memory region to set up the ioeventfd
+ * @offset: offset within the memory region
+ * @size: the size of the region
+ * @fd_index: the index of the fd from the subregion
```

Comment:
this should just be the fd, not the index - how could a client know the index?

Reply:
I don't think we can just use the fd because we never directly return the fd. When we use get_region_io_fds it returns a separate list of fds from the socket which don't have to match the exact number that was used to create the ioeventfds on the server. As such I don't think we can just send the fds as we can never directly get the exact same fds from the server for comparison. As such we would have to pass the index which we get as a return for the sub regions from get_region_io_fds.

Reply:
get_region_io_fds is for the client.

What I'm saying is, when the API consumer (the device impl like SPDK) wants to delete an ioeventfd, then it knows the fd it's for, along with its identifying information. It doesn't know anything about `fd_index` - that's a protocol-level thing to do with the client.

Now, there's an obvious issue here in that we have no way to tell the *client* that an ioeventfd got deleted. So it's really of limited use. In fact, I'd be inclined not to have it at all the more I think about it, at least until we find a need. So, in terms of "less code", shall we remove all the ioeventfd_delete code?

Apart from the clean up in vfu_destroy_ctx of course.

Reply:
Yes I agree with the idea to remove it due to the lack of notification to the client as that is certain to cause a problem with out extreme care being taken.